source/agora/utils/InetUtils.d

```diff
@@ -174,6 +186,135 @@ version (Windows)
+
+        For example:
+            ::5:6:7:8      => 0:0:0:0:5:6:7:8
+            1:2:3::7:8     => 1:2:3:0:0:0:7:8
+            ::             => 0:0:0:0:0:0:0:0
+
+        Params:
+            ip = compressed IPv6 address
+
+        Returns:
+            uncompressed representation of IPv6 address
+
+    ***************************************************************************/
+
+    public static string convertCompressedIPv6ToUncompressed (string ip) @safe pure
+    {
+        auto parts = ip.split(":");
+        const first_empty_ind = parts.countUntil("");
+
+        // filter out all the empty parts, except the first one
+        auto res = zip(iota(parts.length), parts).filter!((tup) => tup[0] == first_empty_ind || !tup[1].empty()).
+                   map!(tup => tup[1]).array();
+
+        // fill in the empty part with the needed zeros
+        return res.map!(part => (part.empty() ? replicate(["0"],8 - res.length + 1) : [part]))
+                    .joiner.array.join(":");
+    }
+
+    unittest
+    {
+        assert(convertCompressedIPv6ToUncompressed("1:2:3:4:5:6:7:8") == "1:2:3:4:5:6:7:8");
+        assert(convertCompressedIPv6ToUncompressed("::5:6:7:8") == "0:0:0:0:5:6:7:8");
+        assert(convertCompressedIPv6ToUncompressed("::8") == "0:0:0:0:0:0:0:8");
+        assert(convertCompressedIPv6ToUncompressed("::5:0:0:0") == "0:0:0:0:5:0:0:0");
+        assert(convertCompressedIPv6ToUncompressed("1:2:3:4:5:6:7::") == "1:2:3:4:5:6:7:0");
+        assert(convertCompressedIPv6ToUncompressed("1::") == "1:0:0:0:0:0:0:0");
+        assert(convertCompressedIPv6ToUncompressed("1:2:3::7:8") == "1:2:3:0:0:0:7:8");
+        assert(convertCompressedIPv6ToUncompressed("::") == "0:0:0:0:0:0:0:0");
+    }
+
+    ///
+    public alias HostPortTup = Tuple!(string, "host", string, "port", HostType, "type");
+
+    /***************************************************************************
+
+        Extracts host and port from a URL and determine the host type
+
+        Host can be DNS host name, IPv6 or IPv4 address. It is important to note
+        that the return value of this method is only correct if the passed in
+        `url` parameter is a well-formed URL. In case of an ill-formed URL the
+        return value of this function should be ignored. The only exception
+        to this rule, is when the returned `HostType` is `Unknown`, in which case
+        the URL is definitely not a DNS hostname, IPv6 or IPv4 address.
+
+        Example URLs are
+            https://www.bosagora.io:1234/data
+            http://[1:2::3]/data
+
+        Params:
+          url = the URL from which we would like to extract the host and port
+
+        Returns: `HostPortTup` containing the extracted host, port and host type
+
+    ****************************************************************************/
+
+    public static HostPortTup extractHostAndPort (string url) @safe
```

Comment:
Couldn't we just use Vibe.d for this ?

Reply:
I looked at the URL class in vibe.d, but it seems very bare with significantly less validation that we have here. It also lacks classifying whether the 'host' type is ipv4, ipv6, or domain name and we need that information.